arborist/server.go

```diff
@@ -280,27 +280,66 @@ func (server *Server) handleAuthMappingPOST(w http.ResponseWriter, r *http.Reque
 		Username string `json:"username"`
 		ClientID string  `json:"clientID"`
 	}{}
-	err := json.Unmarshal(body, &requestBody)
-	if err != nil {
-		msg := fmt.Sprintf("could not parse JSON: %s", err.Error())
-		server.logger.Info("tried to handle auth mapping request but input was invalid: %s", msg)
-		errResponse = newErrorResponse(msg, 400, nil)
+
+	// Try to get username or clientID from the JWT.
+	username := ""
+	clientID := ""
+	if authHeader := r.Header.Get("Authorization"); authHeader != "" {
+		server.logger.Info("Attempting to get username or clientID from jwt...")
+
+		userJWT := strings.TrimPrefix(authHeader, "Bearer ")
+		userJWT = strings.TrimPrefix(userJWT, "bearer ")
+		scopes := []string{"openid"}
+		info, err := server.decodeToken(userJWT, scopes)
+		if err != nil {
+			// Return 400 on failure to decode JWT
+			msg := fmt.Sprintf("tried to get username/clientID from jwt, but jwt decode failed: %s", err.Error())
+			server.logger.Info(msg)
+			errResponse = newErrorResponse(msg, 400, nil)
+		} else {
+			username = info.username
+			clientID = info.clientID
+			if username != "" {
+				server.logger.Info("found username in jwt: %s", username)
+			}
+			if clientID != "" {
+				server.logger.Info("found clientID in jwt: %s", clientID)
+			}
+		}
 	}
-	if (requestBody.Username == "") == (requestBody.ClientID == "") {
-		msg := "must specify exactly one of `username` or `clientID`"
-		server.logger.Info(msg)
-		errResponse = newErrorResponse(msg, 400, nil)
+
+	// If they are not present in the token fallback on the request body
+	if (username == "") && (clientID == "") {
```

Comment:
i would be more comfortable with an `else` or `if authHeader not provided` statement here - `(username == "") && (clientID == "")` could happen in unexpected ways